File: fingerprint-sdk/fingerprint_sdk/ios/Classes/Analytics/DeviceProfiler.swift
import Foundation
import os.log

/// Profiles device capability and returns an adaptive performance tier.
enum DeviceProfiler {

    enum Tier { case high, mid, low }

    struct DeviceProfile {
        let tier: Tier
        let targetWidth: Int
        let targetHeight: Int
        let framesPerFinger: Int
        let enableDFTLiveness: Bool
        let enableSkeletonize: Bool
    }

    static func profile() -> DeviceProfile {
        let ramGb   = Double(ProcessInfo.processInfo.physicalMemory) / 1_073_741_824.0
        let cores   = ProcessInfo.processInfo.processorCount
        let iosVer  = ProcessInfo.processInfo.operatingSystemVersion.majorVersion

        let tier: Tier
        if ramGb >= 3.0 && cores >= 6 { tier = .high }
        else if ramGb >= 1.5 && cores >= 4 { tier = .mid }
        else { tier = .low }

        os_log("DeviceProfiler: RAM=%.1fGB cores=%d iOS=%d → tier=%{public}s",
               log: OSLog(subsystem: "com.yellowsense.fingerprint_sdk", category: "Device"),
               type: .debug, ramGb, cores, iosVer, "\(tier)")

        switch tier {
        case .high:
            return DeviceProfile(tier: .high, targetWidth: 1280, targetHeight: 720,
                                 framesPerFinger: 5, enableDFTLiveness: true, enableSkeletonize: true)
        case .mid:
            return DeviceProfile(tier: .mid, targetWidth: 960, targetHeight: 540,
                                 framesPerFinger: 3, enableDFTLiveness: true, enableSkeletonize: true)
        case .low:
            return DeviceProfile(tier: .low, targetWidth: 640, targetHeight: 480,
                                 framesPerFinger: 3, enableDFTLiveness: false, enableSkeletonize: false)
        }
    }
}

File: fingerprint-sdk/fingerprint_sdk/android/src/main/kotlin/com/yellowsense/fingerprint_sdk/analytics/DeviceProfiler.kt
package com.yellowsense.fingerprint_sdk.analytics

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import android.util.Log

/**
 * Profiles device capability at SDK init time and returns a performance tier.
 * Low-end devices get reduced resolution and simplified processing.
 */
object DeviceProfiler {

    enum class Tier { HIGH, MID, LOW }

    data class DeviceProfile(
        val tier: Tier,
        val targetResolutionWidth: Int,
        val targetResolutionHeight: Int,
        val framesPerFinger: Int,       // how many frames to accumulate before picking best
        val enableDFTLiveness: Boolean, // DFT is expensive; skip on LOW tier
        val enableSkeletonize: Boolean  // skeletonisation is expensive; skip on LOW tier
    )

    fun profile(context: Context): DeviceProfile {
        val ramMb   = totalRamMb(context)
        val cores   = Runtime.getRuntime().availableProcessors()
        val apiLevel = Build.VERSION.SDK_INT

        val tier = when {
            ramMb >= 3000 && cores >= 6 -> Tier.HIGH
            ramMb >= 1500 && cores >= 4 -> Tier.MID
            else                        -> Tier.LOW
        }

        Log.d("DeviceProfiler", "RAM=${ramMb}MB cores=$cores api=$apiLevel → tier=$tier")

        return when (tier) {
            Tier.HIGH -> DeviceProfile(
                tier                  = Tier.HIGH,
                targetResolutionWidth  = 1280,
                targetResolutionHeight = 720,
                framesPerFinger        = 5,
                enableDFTLiveness      = true,
                enableSkeletonize      = true
            )
            Tier.MID -> DeviceProfile(
                tier                  = Tier.MID,
                targetResolutionWidth  = 960,
                targetResolutionHeight = 540,
                framesPerFinger        = 3,
                enableDFTLiveness      = true,
                enableSkeletonize      = true
            )
            Tier.LOW -> DeviceProfile(
                tier                  = Tier.LOW,
                targetResolutionWidth  = 640,
                targetResolutionHeight = 480,
                framesPerFinger        = 3,
                enableDFTLiveness      = false,  // skip expensive DFT
                enableSkeletonize      = false   // use Canny ridges directly
            )
        }
    }

    private fun totalRamMb(context: Context): Long {
        val am   = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val info = ActivityManager.MemoryInfo()
        am.getMemoryInfo(info)
        return info.totalMem / (1024 * 1024)
    }
}
